[channel-namming] - supplement recommendations

camilamacedo86 opened the pull request:
**Description**

We are proposing some additionals on its [documentation](https://github.com/operator-framework/olm-docs/pull/232/files). **What is the delta?**

- Clarifies why the terminologies tech-preview or preview is not in the convention
- We added as the prefix of the operator visions the `v` (i.e. stable-v2, stable-v2.3 ) to better clarifies it because of the scenario where the channel name has ONLY the Operato major version.
- Also, we added a new section `Recommended upgrade path` to clarifies and exemplifies :

> Within a channel each patch release should be directly upgradable to the HEAD of channel. Use skips or skipRange to provide this behaviour. (i.e. skipRange to publish a patch for `3.6.z`under the`stable-v3.6` would mean setting the skipRange to be `>= 3.5.z. < 3.6.z`, where `3.5.z` represents the oldest `3.5.z` version for which you can provide direct upgrade support to `3.6.z` latest)

Nit: We added the links to the respective docs as well.

camilamacedo86 commented:
c/c @dmesser

bparees commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -175,9 +198,52 @@ candidate/fast/stable to denote the maturity level of the operator.  Using all
+all versions which are provided via the `stable-3.5` channel. 
+
+So that, you can ensure that your users can move forward from any version that they have installed promoted via the 
+`stable-3.5` channel to the first version of the channel `stable-3.6`. (e.g. `skipRange: ">=3.5.0 <3.6.0"`)
```
Caveat to call out here is this means that the operator author must be comfortable that any 3.6.z version can upgrade from any 3.5.z version.

bparees commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -175,9 +198,52 @@ candidate/fast/stable to denote the maturity level of the operator.  Using all
+#### Example
+
+If you have published solutions on the channel `stable-3.5`
+then, you would like to begin to distribute Operator bundle versions `3.6.x`
```
```suggestion
then, you would like to begin to distribute Operator bundle versions `3.6.z`
```

(versioning is X.Y.Z)

joelanford commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -136,21 +146,34 @@ can experiment with that fast version, but the stable version for example
+issues on their cluster because they know that none of the solutions which will be
+distributed within contains breaking changes:
+
+| Channels for Operator version(s) 2.x | 
+| :------------- | 
+| candidate-2.x |
+| fast-2.x | 
+| stable-2.x |
```
I think I'd leave off the `.x` suffix to align with major/minor channel names and make it more clear that these channels track all releases within a major version.

joelanford commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -175,9 +198,52 @@ candidate/fast/stable to denote the maturity level of the operator.  Using all
+
+If you have the channel `stable-3.7` where the head of channel is `3.7.10` and then, you will provide a new patch release
+with a bug fix using the Operator bundle version `3.7.11` then, ensure that you configure your new publication to skip all 
+that was published in under the channel `stable-3.7` (e.g. skipRange: `">= 3.7.0 < 3.7.11"` replaces: `3.7.10` ).  
```
+1 Ben

The problem is that most readers are likely still using the sqlite imperative workflow, so its very difficult (maybe impossible) to track a moving target from the previous y stream.

This would be very easy to implement in FBC, especially with a veneer that implements all the logic that's described.

bparees commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -96,25 +96,35 @@ spec:
 * You could include or have an operand version in your channel name to advertise to consumers the version of operand they can subscribe to.
 * If you do choose to include some version in your channel name, it is important to include an additional identifier, to clarify what that the version number is referring to. A version number could equally be referring to product version (operand version), or operator version - these two don't always match: the operator itself can have different versioning than the product it is managing. 
-
+* It is recommended to use the major versions of your Operator releases in the channel names at least. So that cluster admins can better plan when they can begin to consume newer versions of your Operators, which contains breaking changes and avoid workflow issues. (e.g. `stable-2.x`)
```
```suggestion
* It is recommended to use the major versions of your Operator releases in the channel names at least. So that cluster admins can better plan when they can begin to consume newer versions of your Operators, which contains breaking changes and avoid workflow issues. (e.g. `stable-2`)
```

(if you're going to align w/ the examples below.  Though i have to say, it's not super obvious that "2" is a version number in that name.  @joelanford i think you proposed `stable-2` instead of `stable-2.x`, do you have thoughts here?  maybe it's not worth worrying about, or maybe we recommend `stable-v2`?  I don't know how many products will do a major version only name, vs "stable-2.1" which is less of an issue))

joelanford replied:
To me, its obvious if there are y-named channels there for context `stable-2`, `stable-2.0`, `stable-2.1`, etc. That's what I've typically seen with semver-based docker image versioning (e.g. [nginx](https://hub.docker.com/_/nginx) and [debian](https://hub.docker.com/_/debian)

I'm not opposed to a `v` prefix to make that more obvious though.

bparees replied:
> To me, its obvious if there are y-named channels there for context stable-2, stable-2.0, stable-2.1

if, yes.  I don't know how many teams are going to maintain *both* "major-version" channels (foo-X) *and* "minor version" channels(foo-X.Y) though, so i don't know if you can really assume that there will be a "foo-X.Y" that lends context to a "foo-X"

camilamacedo86 replied:
I am OK with we add the prefix`v`.
Indeed I think that is a good idea but then that means adding it for all scenarios.

bparees commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -136,21 +146,36 @@ can experiment with that fast version, but the stable version for example
+issues on their cluster because they know that none of the solutions which will be
+distributed within contains breaking changes:
+
+| Channels for Operator version(s) 2.z | 
```
```suggestion
| Channels for Operator version(s) 2.x | 
```

bparees commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -175,9 +200,80 @@ candidate/fast/stable to denote the maturity level of the operator.  Using all
+## Recommended upgrade path for patch release (bug fixes)
+
+Within a channel each patch release should be directly upgradable to the `head channel` from the latest version shipped 
+in the minor channel behind. Use skips or skipRange to provide this behaviour. (i.e. using the skipRange in the 
```
```suggestion
in the previous minor version channel. Use skips or skipRange to provide this behaviour. (i.e. using the skipRange in the 
```

bparees commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -175,9 +200,80 @@ candidate/fast/stable to denote the maturity level of the operator.  Using all
+
+Within a channel each patch release should be directly upgradable to the `head channel` from the latest version shipped 
+in the minor channel behind. Use skips or skipRange to provide this behaviour. (i.e. using the skipRange in the 
+`stable-3.6` patch for `3.5.<latest>: >= 3.5.<first z of the minor channel>. < 3.6.<new pacth release>`)
```
```suggestion
`stable-3.6` patch for `3.5.<latest>: >= 3.5.<first z of the minor channel>. < 3.6.<new patch release>`)
```

aside from the typo, i can't understand what this is saying.

should it be this instead:
`stable-3.6` patch for `3.6.<latest>` would mean setting the skipRange to be `>= 3.5.<first z of the minor channel>. < 3.6.<new patch release>`, where 3.5.z represents the oldest 3.5.z version for which you can provide direct upgrade support to 3.6.latest)

bparees commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -175,9 +200,80 @@ candidate/fast/stable to denote the maturity level of the operator.  Using all
+- Operator bundles versions from `3.6.0` to `3.6.29` under the `stable-3.6` channel
+
+Then, in this case, your newer patch release `3.6.30` would be configured with `skipRange: >=3.5.25 < 3.6.30` in
+order to only supports upgrading to the newest `3.6.z` from a pretty recent `3.5.z`
```
They shouldn't necessarily use 3.5.25, they should use the oldest 3.5.z version that they can support upgrades from.

they can use 3.5.25 if, for some reason, they don't think upgrades from something older like 3.5.24 (or more likely, 3.5.0) to 3.6.latest will work.  But it the ideal situation is that they do allow upgrades from 3.5.0 to 3.6.latest (i.e. `skipRange: >=3.5.0 < 3.6.30`).  The ability to set it to 3.5.25 (or 3.5.17, or something else) instead of 3.5.0 is a mechanism to constrain the support matrix *when necessary*, it's not the default recommendation.

bparees commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -175,9 +200,80 @@ candidate/fast/stable to denote the maturity level of the operator.  Using all
+
+**Note** If you would like to ensure that the users of your Operator still able to install the Operator bundle version
+`3.6.29` by using the option `startingCSV` added manually in the subscription then, you will need to also use the `replaces: 3.6.29`
+for this Operator bundle does not get pruned from the index catalog via the OLM resolver.
```
```suggestion
to ensure the older Operator bundle does not get pruned from the index catalog via the OLM resolver.
```

bparees commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -175,9 +200,80 @@ candidate/fast/stable to denote the maturity level of the operator.  Using all
+
+However, **attention** that means you would no longer able to provide any patch release to the minor channel `stable-3.6`.    
+Note that if you publish `3.6.latest+1` when this version comes out, and your users upgrades to it, they will have no way to
+upgrade from `3.6.latest+1` to any solution published under the channel `stable-3.7`.
```
suggest this instead for the above paragraph:


However, that means that if/when you publish a `3.6.latest+1` and a user upgrades to it, they will not
have a path to `3.7` because `3.6.latest+1` is not included in the skipRange defined by any existing `3.7.z` bundle.  You would have to then publish a new `3.7.z` bundle that expands the skipRange to include 3.6.latest+1 before users would be able to upgrade to a 3.7.z version.

bparees replied:
(this is a restriction that becomes less problematic with FBC when operator authors are able to update the skipRange of a previously published bundle, so they could extend the skipRange of 3.7.11 to include 3.6.latest+1, but of course we aren't there yet)

camilamacedo86 replied:
I think a follow up here will be to change this doc for FBC or tweak it. 
IMO I do not think that we should only FBC instructions any way since it will be optional in the future for the authors and many of them might never adopt it.

bparees replied:
>  I do not think that we should only FBC instructions any way since it will be optional in the future for the authors and many of them might never adopt it.

at a minimum i agree it's probably too early to do so.

joelanford replied:
this is reminiscent of the discussion we had many months ago when we replaced some imperative `opm registry add` docs with the new FBC docs.

the main branch of the upstream docs site focuses on FBC, not the deprecated sqlite workflows.

joelanford replied:
to reframe this as a question, why should we recommend that upstream users keep using deprecated workflows when FBC exists and helps solve a bunch of these problems?

if there are pipelines that have not adopted FBC, those pipelines likely need to have documentation about their requirements and assumptions.

IMO the upstream OLM docs site should help catalog maintainers and pipeline builders understand how to get off the deprecated path and onto FBC.

camilamacedo86 replied:
Closing this one since we spoke shows that we agree that its goal is only to add what is missing we probably want to work in the future to see how we can shape it for FBC.

bparees commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -175,9 +200,80 @@ candidate/fast/stable to denote the maturity level of the operator.  Using all
+
+Because of this, it is recommended configure your bundles versions to allow cross-channel updates at least 
+between two consecutive major/minor versions (e.g. by having installed the head of the channel `stable-2.9` 
+your users can get installed the first version under `stable-2.10`)
```
this whole section seems redundant to the previous section(` Recommended upgrade path for patch release (bug fixes)`) and the recommendation is worse (because it recommends upgrade paths which take the user through the X.Y.0 version of a minor stream, instead of directly to X.Y.latest).

everything recommended in the previous section already allows for a user to move from "stable-2.9" to "stable-2.10" if it is done correctly.  (i.e. the previous section covers both patch upgrades *and* minor version upgrades).

I suggest dropping this section and simply retitling the previous section to clarify that it's upgrade graph structure recommendations generally.

camilamacedo86 commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -96,25 +96,35 @@ spec:
 * You could include or have an operand version in your channel name to advertise to consumers the version of operand they can subscribe to.
 * If you do choose to include some version in your channel name, it is important to include an additional identifier, to clarify what that the version number is referring to. A version number could equally be referring to product version (operand version), or operator version - these two don't always match: the operator itself can have different versioning than the product it is managing. 
-
+* It is recommended to use the major versions of your Operator releases in the channel names at least. So that cluster admins can better plan when they can begin to consume newer versions of your Operators, which contains breaking changes and avoid workflow issues. (e.g. `stable-2.x`)
```
```suggestion
* It is recommended to use the major versions of your Operator releases in the channel names at least. So that cluster admins can better plan when they can begin to consume newer versions of your Operators, which contains breaking changes and avoid workflow issues. (e.g. `stable-v2`)
```

bparees commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -175,9 +200,60 @@ candidate/fast/stable to denote the maturity level of the operator.  Using all
+
+You could ensure that your users can move forward easier to the latest version since in this range `">= 3.7.0 < 3.7.11"`
+as they will be able to install the new `3.7.11` release from  `3.6.z <latest release on 3.6 channel>` directly 
+instead of get installed first the Operator version `3.7.0` which we know that have bug fixes in place already.
```
```suggestion
instead of having to upgrade through Operator version `3.7.0` which may contain bugs that are already fixed in both the 3.6.z version they had installed, and the 3.7.11 version they are trying to get to.
```

bparees commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -175,9 +200,60 @@ candidate/fast/stable to denote the maturity level of the operator.  Using all
+as they will be able to install the new `3.7.11` release from  `3.6.z <latest release on 3.6 channel>` directly 
+instead of get installed first the Operator version `3.7.0` which we know that have bug fixes in place already.
+
+However, **attention** that means you would no longer able to provide any patch release to the minor channel `stable-v3.6`.    
```
```suggestion
However, this approach has negative implications when you provide your next patch release to the minor channel `stable-v3.6`.    
```

bparees commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -175,9 +200,60 @@ candidate/fast/stable to denote the maturity level of the operator.  Using all
+
+However, **attention** that means you would no longer able to provide any patch release to the minor channel `stable-v3.6`.    
+Note that if you publish `3.6.latest+1` when this version comes out, and your users upgrades to it, they will have no way to
+upgrade from `3.6.latest+1` to any solution published under the channel `stable-v3.7`.
```
```suggestion
upgrade from `3.6.latest+1` to any solution published under the channel `stable-v3.7`, until you publish a new 3.7.z version that replaces 3.6.latest+1.
```

bparees reviewed (approved):
one nit, otherwise lgtm.

camilamacedo86 commented:
c/c @anik120

camilamacedo86 commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -96,25 +96,35 @@ spec:
 ## NAMING
 
-Channel names are used to imply what form of upgrade you want to offer for your operator.  For example, you might have an operator that has a candidate or alpha version which is not supported as well as a version where support is offered.
+Channel names are used to imply what form of upgrade you want to offer for your operator.  For example, you might have an operator that has a candidate release version which is not supported as well as a version where support is offered.
+
+Our motivation with this convention is to encourage a better user experience for OLM users to have an intuitive 
+understanding of the level of maturity and supportability of the solutions that will get installed. It can get very complex
+for cluster admins and Operator consumers in general when each solution distribute in the OLM catalog adopts different names
+or different meanings for the same terminologies.
```
```suggestion
Channel names are used to imply what form of upgrade you want to offer for your operator. For example, you might have an operator with a candidate release version that is not supported and a version where support is offered.
		

Our motivation with this convention is to encourage a better user experience for OLM users to have an intuitive 
understanding of the level of maturity and supportability of the solutions that will get installed. 
It can get very complex for cluster admins and Operator consumers in general when each solution distributed in the OLM catalog adopts different names or different meanings for the same terminologies.
```

camilamacedo86 commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -96,25 +96,35 @@ spec:
 * Channel names are chosen by operator authors as they see fit to meet their upgrade strategies.
+* Channel names are related to the release strategy and not to specific features. ( e.g. we can have an Operator bundle release which is supported and ought to be under the stable channel but with a "tech-preview" and not supported feature. ).
+
+> Note: If your new operator bundle release contains an API(CRD), which represents new experimental features and not supported (like "tech-preview"), then
+the terminology recommended to being adopted would be in the API version (e.i. group: my.example.com  kind: Backup version: v1alpha1) and **not** in the channel naming. The recommendation for this scenario is to follow up the [Kubernetes API versioning](https://kubernetes.io/docs/reference/using-api/#api-versioning).
```
```suggestion
> **Note:** If your new operator bundle release contains an API(CRD), which represents new experimental features and is not supported (like "tech-preview"), then the terminology recommended to adopt would be in the 
API version (e.i. group: my.example.com kind: Backup version: v1alpha1) and **not** in the channel naming. The recommendation for this scenario is to follow up the [Kubernetes API versioning](https://kubernetes.io/docs/reference/using-api/#api-versioning).
```

fgiloux commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -136,33 +146,50 @@ can experiment with that fast version, but the stable version for example
+| fast-v2 | 
+| stable-v2 |
+
+**Attention:** The versions on the channel names are related to the Operator version and not its Operand.
```
Nit: I would add "s" to "Operand"

fgiloux commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -136,33 +146,50 @@ can experiment with that fast version, but the stable version for example
+| fast-v1.3 | fast-v2.4 |
+| stable-v1.3 | stable-v2.4 | 
+
+**Attention:** The versions on the channel names are related to the Operator version and not its Operand.
```
Nit: I would add "s" to "Operand"

fgiloux commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -175,9 +202,62 @@ candidate/fast/stable to denote the maturity level of the operator.  Using all
+# UPGRADES 
+
+OLM provides a variety of ways to specify updates between operator versions. Before
+continue in the following recommendations, make sure you understand its options by checking [Creating an update graph with OLM][upgrade-graph].
```
Proposing "Before we continue with the following recommendations, make sure you understand the options by checking [Creating an update graph with OLM][upgrade-graph]."

fgiloux commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -175,9 +202,62 @@ candidate/fast/stable to denote the maturity level of the operator.  Using all
+using this example from an Operator version like `3.5.25` _(or `3.5.17`, or something else)_ 
+instead of `3.5.0` is a mechanism to constrain the support matrix when necessary, it's not the default recommendation.
+
+**Note** If you would like to ensure that the users of your Operator still able to install the Operator bundle version
```
Wording; "that the users of your Operator still able" -> "that the users of your Operator are still able"

fgiloux commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -175,9 +202,62 @@ candidate/fast/stable to denote the maturity level of the operator.  Using all
+`3.6.29` by using the option `startingCSV` added manually in the subscription then, you will need to also use the `replaces: 3.6.29`
+to ensure the older Operator bundle does not get pruned from the index catalog via the OLM resolver.
+
+#### Attention (Be aware of the following example scenario)
```
Proposing: "Be aware of the following scenario"

fgiloux commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -175,9 +202,62 @@ candidate/fast/stable to denote the maturity level of the operator.  Using all
+as they will be able to install the new `3.7.11` release from  `3.6.z <latest release on 3.6 channel>` directly
+instead of having to upgrade through Operator version `3.7.0` which may contain bugs that are already fixed in both the `3.6.z` version 
+they had installed, and the `3.7.11` version they are trying to get to.
+
```
Proposing: "By doing so you could ensure that your users can more easily upgrade to the latest version since they will be able to install the new `3.7.11` release directly from `3.6.z <latest release on 3.6 channel>` instead of having to upgrade through Operator version `3.7.0` which may contain bugs that are already fixed in both the `3.6.z` version they had installed, and the `3.7.11` version they are moving to."

fgiloux commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -175,9 +202,62 @@ candidate/fast/stable to denote the maturity level of the operator.  Using all
+they had installed, and the `3.7.11` version they are trying to get to.
+
+However, this approach has negative implications when you provide your next patch release to the minor channel `stable-v3.6`.  
+Note that if you publish `3.6.latest+1` when this version comes out, and your users upgrades to it, they will have no way to
```
Nit: "and your users upgrade to it"

fgiloux commented:
/lgtm
my comments have been addressed.
In regards to the discussion legacy vs FBC I believe that they will cohabit for some more time and there is value in having both documented. Community operators and their catalogs may move quicker than operators and their catalogs built within enterprise having strict and time consuming validation processes, slow moving infrastructures. We still need to keep them onboard.

anik120 commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -96,25 +96,37 @@ spec:
 ## NAMING
 
-Channel names are used to imply what form of upgrade you want to offer for your operator.  For example, you might have an operator that has a candidate or alpha version which is not supported as well as a version where support is offered.
+Channel names are used to imply what form of upgrade you want to offer for your operator. For example, you might have an operator with a candidate release version that is not supported and a version where support is offered.
```
Do you think it'll be helpful to clarify this point a little more with examples? 
```suggestion
Channel names should imply the nature of official support the user can expect when subscribed to that channel for your operator. For example, the channel name `stable-v2` implies that subscribing to this channel will install an operator with a major version `v2` that you can expect official support for. The channel name `candidate-v2` implies that the user will be installing a candidate release version that does not have official supported.
```

camilamacedo86 replied:
The examples are added, I mean, they exist already in the doc

anik120 commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -96,25 +96,37 @@ spec:
+
+Our motivation with this convention is to encourage a better user experience for OLM users to have an intuitive 
+understanding of the level of maturity and supportability of the solutions that will get installed.
+It can get very confusing for cluster admins and Operator consumers in general when each solution distributed in the OLM catalog adopts different names or different meanings for the same terminologies.
```
"It can be very confusing for cluster admins and Operator consumers in general when each solution distributed in the OLM catalog adopts different names or different meaning for the same terminologies" 

Is this sentence really required in this doc? We know it's out motivation to write this entire channel naming convention doc, but just reading the sentence it looks like it has the potential to throw off users with words like "solution", "terminologies" etc. We know what they mean to us who's been looking at these "solutions" and "terminoliges" for a long time now but a new operator author will probably be very confused with this sentence (and thereby increase the cognitive overload needed for reading this doc)

camilamacedo86 replied:
One of the goals is to clarify why we need the convention.
Besides, it shows obvious for us that have not been obvious to everybody.

anik120 commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -96,25 +96,37 @@ spec:
 ### Naming Conventions
 
 * Channel names are chosen by operator authors as they see fit to meet their upgrade strategies.
+* Channel names are related to the release strategy and not to specific features. ( e.g. we can have an Operator bundle release which is supported and ought to be under the stable channel but with a "tech-preview" and not supported feature. ).
```
```suggestion
* Channel names should communicate the release strategy, official support level and the level of maturity, but **not** specific features. ( e.g. a channel with Operator versions which has official support ought to be have the word `stable` in the channel name, whereas a channel with Operator versions with a non-supported feature should have the phrase `tech-preview`in it's name. ).
```

anik120 commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -96,25 +96,37 @@ spec:
 * Channel names are chosen by operator authors as they see fit to meet their upgrade strategies.
+* Channel names are related to the release strategy and not to specific features. ( e.g. we can have an Operator bundle release which is supported and ought to be under the stable channel but with a "tech-preview" and not supported feature. ).
+
+> **Note:** If your new operator bundle release contains an API(CRD), which represents new experimental features and
```
```suggestion
> **Note:** If your new operator version release contains an API(CRD), which represents new experimental features and
```

anik120 commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -136,33 +148,48 @@ can experiment with that fast version, but the stable version for example
+
+#### Example 3 (Recommended Option is the most common scenarios)
+
+An operator wants to be supported at various operator major/minor versions.  
```
```suggestion
An operator should be supported at various operator major/minor versions.  
```

anik120 commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -136,33 +148,48 @@ can experiment with that fast version, but the stable version for example
+#### Example 3 (Recommended Option is the most common scenarios)
+
+An operator wants to be supported at various operator major/minor versions.  
+For example, you might have an operator version at `1.3` and also at `2.4` that 
```
```suggestion
For example, you might have a `v1.3` and also a `v2.4` releases of your operator that 
```

anik120 commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -136,33 +148,48 @@ can experiment with that fast version, but the stable version for example
+An operator wants to be supported at various operator major/minor versions.  
+For example, you might have an operator version at `1.3` and also at `2.4` that 
+you need or want to offer support for at the same time.  However, you might 
+not want to have OLM upgrade users to the `2.4` Operator but instead keep them 
```
```suggestion
not want to have OLM upgrade users to the `v2.4` Operator but instead keep them 
```

anik120 commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -175,9 +202,62 @@ candidate/fast/stable to denote the maturity level of the operator.  Using all
+## Recommended upgrade path
+
+Within a channel each patch release should be directly upgradable to the [HEAD of channel](/docs/glossary/#channel-head). 
+Use skips or skipRange to provide this behaviour. (i.e. if you followed the above `Example 3` that means 
```
We should probably link to the `skips` and `skipsRange` sections of the doc site from here

camilamacedo86 replied:
Usually, I would agree with. But then it can bring confusion Why?
skip/skipRange/replaces configured via CSV still supported and valid indeed with FBC format. 
However, our latest version of this doc ONLY explains how to do it with the new FBC options and we remove completed this explanation about how to do it using the CSV.

anik120 commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -175,9 +202,62 @@ candidate/fast/stable to denote the maturity level of the operator.  Using all
+Let's imagine that you will publish the Operator bundle `3.6.30` (patch release under the channel `stable-v3.6`) and
+that you have published so far:
+
+- Operator bundles versions from `3.5.0` to `3.5.25` under the `stable-v3.5` channel
```
```suggestion
- Operator bundles versions from `v3.5.0` to `v3.5.25` under the `stable-v3.5` channel
```

This is a nit, for all the versions quoted in the rest of the section too, just to make it look a little more clean by following the semver convention

anik120 commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -175,9 +202,62 @@ candidate/fast/stable to denote the maturity level of the operator.  Using all
+
+#### Attention (Be aware of the following scenario)
+
+If you have the channel `stable-v3.7` where the head of channel is `3.7.10` and then, you provide a new patch release
```
```suggestion
If you have the channel `stable-v3.7` where the head of channel is `v3.7.10`, and you have a new patch release
```

anik120 commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -175,9 +202,62 @@ candidate/fast/stable to denote the maturity level of the operator.  Using all
+#### Attention (Be aware of the following scenario)
+
+If you have the channel `stable-v3.7` where the head of channel is `3.7.10` and then, you provide a new patch release
+with a bug fix using the Operator bundle version `3.7.11` then, imagine that you configure your new publication to skip all
```
```suggestion
with a bug fix using the Operator bundle version `v3.7.11` , and you configure `v3.7.11` to skip all
```

anik120 commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -175,9 +202,62 @@ candidate/fast/stable to denote the maturity level of the operator.  Using all
+
+If you have the channel `stable-v3.7` where the head of channel is `3.7.10` and then, you provide a new patch release
+with a bug fix using the Operator bundle version `3.7.11` then, imagine that you configure your new publication to skip all
+that was published under the channel `stable-v3.7` (e.g. skipRange: `">= 3.7.0 < 3.7.11"` replaces: `3.6.z <latest release on 3.6 channel>` ).
```
```suggestion
operator bundles published in the channel `stable-v3.7` (e.g.`v3.7.11` mentions  skipRange: `">= 3.7.0 < 3.7.11"` replaces: `3.6.z <latest release on 3.6 channel>` ).
```

anik120 commented on content/en/docs/best-practices/channel-naming.md:
```diff
@@ -175,9 +202,62 @@ candidate/fast/stable to denote the maturity level of the operator.  Using all
+with a bug fix using the Operator bundle version `3.7.11` then, imagine that you configure your new publication to skip all
+that was published under the channel `stable-v3.7` (e.g. skipRange: `">= 3.7.0 < 3.7.11"` replaces: `3.6.z <latest release on 3.6 channel>` ).
+
+By doing so you could ensure that your users can more easily upgrade to the latest version since they will 
```
```suggestion
, you could ensure that your users can more easily upgrade to the latest version since they will 
```

anik120 reviewed (changes requested):
Thanks @camilamacedo86 for the PR. Left some suggestions.

camilamacedo86 commented:
Hi, [@anik120](https://coreos.slack.com/team/UF9GJC7EK) really thank you for your help.
I addressed all comments and mainly accepted all (99%) of your suggestions.
I hope that it is good enough to get merged now. Let us know. 





[9:41](https://coreos.slack.com/archives/D0108CA2L4U/p1649925670757859)
I hope that we can move forward with it now.
[9:41](https://coreos.slack.com/archives/D0108CA2L4U/p1649925681561979)
Again, thx a lot for your time.

cdjohnson commented:
@camilamacedo86 I do have some concerns, which reflect why we did not adopt this naming strategy.  I apoologize for not commenting earlier, but I wanted some of the internal discussions and noise to die down first....

You can see my research [here](https://hackmd.io/5pjcHqHsQtmiJGd_WLUulA?view) for some backround.

Regarding the suggested channel naming of `<frequency>-<version>`:
1.  Unless your delivery pipeline is really good, in order to promote artificial operator versions to include the promotion logic is very cumbersome.  It didn't work for us at all.  We therefore could not follow this model.
2. It would be worthwhile to highlight that FBC completely solves the promotion problem, making this model totally realistic.
3. It makes more sense to me to flip the channel names to `<version>-<frequency` (e.g. v1-candidate), so like-versions are lexicographically sorted together. 
4. Our customers REALLY want to control the risk of upgrades.  That's why we introduced `vX.Y` and stable-only (no frequency).  With FBC this is opening the door for all sorts of much more interesting patterns, and hence why my hackmd.io doc is there....

camilamacedo86 commented:
Hi @cdjohnson, 

Regards your comment  https://github.com/operator-framework/olm-docs/pull/232#issuecomment-1100098507

It seems like you are trying to add here what would be possible to achieve in the future with FBC.
That is not the goal of this PR and is out of its scope. Note that this document highlighted already the current limitations regards channels promotion which is something that FBC will help you out with.

anik120 commented:
/test/link(pull_request) override since a link in the contribution guideline doc is failing and this PR does not touch that doc, and has gone through enough churn. We can fix that issue in another PR. 

/lgtm
/approve
